src/state/opamSwitchState.ml

```diff
@@ -973,6 +975,19 @@ let universe st
       ~force_dev_deps ~test ~doc ~dev_setup
       ~requested_allpkgs
   in
+  (* We retrieve the previous opam files, package in this set are simulated
+     pins and we don't want the take the new definition into account for
+     solving *)
+  let overwrote_opams, overwrote_unpinned =
+    OpamPackage.Map.fold (fun nv (pinned, opam) (opams, unpinneds) ->
+        OpamPackage.Map.add nv opam opams,
+        if pinned then unpinneds else OpamPackage.Set.add nv unpinneds)
+      st.overwrote_opams OpamPackage.(Map.empty, Set.empty)
+  in
+  let opams =
+    OpamPackage.Map.union (fun _ o -> o) st.opams overwrote_opams
+  in
+  let u_pinned = OpamPinned.packages st -- overwrote_unpinned in
```

Comment:
This is not the most optimal, it can be directly computed during the fold, but i found it more clear that way

The other way:
```suggestion
  let opams, u_pinned =
    OpamPackage.Map.fold (fun nv (was_pinned, opam) (opams, pinned) ->
        OpamPackage.Map.add nv opam opams,
        if was_pinned then pinned else OpamPackage.Set.remove nv pinned)
      st.overwrote_opams (st.opams, OpamPinned.packages st)
  in